article_store/articles/api.py

```diff
@@ -0,0 +1,206 @@
+from django.http.response import HttpResponse
+from django.core.exceptions import ObjectDoesNotExist, ValidationError
+from rest_framework import status, viewsets
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework_xml.renderers import XMLRenderer
+
+from articles.xml_parser import ArticleXMLParser
+
+from articles.models import (
+    Article,
+    ArticleVersion,
+    Content,
+)
+from articles.serializers import (
+    ArticleSerializer,
+    ArticleVersionSerializer,
+)
+
+from articles.utils import parse_accept_language_header
+
+
+class ArticleViewSet(viewsets.ModelViewSet):
+    model = Article
+    queryset = Article.objects.all()
+    serializer_class = ArticleSerializer
+
+
+def article_list_xml_generator(articles: 'QuerySet[Article]') -> Element:
+    """Generate an xml representation of all `Article` objects.
+
+    :param articles: QuerySet[Article]
+    :return: class: `Element`
+    """
+    name_space = {'libero': 'http://libero.pub'}
+
+    root = Element('{%s}articles' % name_space['libero'], nsmap=name_space)
+
+    for article in articles:
+        child = SubElement(root, '{%s}article' % name_space['libero'])
+        child.text = str(article.id)
+
+    return root
+
+
+class ArticleXMLAPIView(APIView):
+    parser_classes = (ArticleXMLParser,)
+    renderer_classes = (XMLRenderer,)
+
+    default_part_name = 'front'
+    latest = 'latest'
+    style_element = '<?xml version="1.0" ?>\n'
+
+    def post(self, request: Request, article_id: str) -> HttpResponse:
+        """
+
+        :param request: class: `Request`
+        :param article_id: str
+        :return: class: `HttpResponse`
+        """
+
+        if not article_id or not Article.id_is_valid(article_id):
+            return Response({'error': 'Please provide a valid article id'}, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        article, created = Article.objects.get_or_create(id=article_id)
+        article_version = ArticleVersion.objects.create(version=article.next_version,
+                                                        article=article)
+
+        for content_item in request.data.get('content-list', []):
+            Content.objects.create(article_version=article_version, **content_item)
+
+        serializer = ArticleVersionSerializer(article_version)
+
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    @staticmethod
+    def put(request: Request, article_id: str, version: str) -> HttpResponse:
+        """Update an `ArticleVersion`
+
+        :param request:
+        :param article_id:
+        :param version:
+        :return: class: `HttpResponse`
+        """
+        if article_id and version:
+            article_version = ArticleVersion.objects.get(version=version, article_id=article_id)
+
+            old_content = Content.objects.filter(article_version=article_version)
+            old_content.delete()
```

Comment:
I think it's fine to delete and re-add, but also by ensuring it's a single database transaction so that clients never see an incomplete version